qemu_build

```diff
@@ -0,0 +1,165 @@
+
+mkdir -p .kvm-images/{amd64,arm64}
+
+if [[ ! -f .kvm-images/amd64/bionic-server-cloudimg-amd64.img && "$PLATFORM" == "amd64" ]]; then
+  curl -SL https://cloud-images.ubuntu.com/bionic/current/bionic-server-cloudimg-amd64.img > .kvm-images/amd64/bionic-server-cloudimg-amd64.img
+fi
+
+if [[ ! -f .kvm-images/arm64/bionic-server-cloudimg-arm64.img && "$PLATFORM" == "arm64" ]]; then
+  curl -SL https://cloud-images.ubuntu.com/bionic/current/bionic-server-cloudimg-arm64.img > .kvm-images/arm64/bionic-server-cloudimg-arm64.img
+  curl -SL https://releases.linaro.org/components/kernel/uefi-linaro/latest/release/qemu64/QEMU_EFI.fd > .kvm-images/arm64/QEMU_EFI.fd
+fi
+
+
+debian_snapshot_id=${3:-}
+
+IMAGE_FILE="build/$DIST/$PLATFORM/instance.qcow2"
+PIDFILE="build/$DIST/$PLATFORM/instance.pid"
+TARGET_FILE="build/$DIST/$PLATFORM/image.tar"
+
+mkdir -p "build/$DIST/$PLATFORM"
+
+qemu-img create -f qcow2 -o backing_file="../../../.kvm-images/$PLATFORM/bionic-server-cloudimg-$PLATFORM.img" "$IMAGE_FILE"
+qemu-img resize "$IMAGE_FILE" 8G
+
+USER_DATA='
+#cloud-config
+users:
+    - name: ubuntu
+      password: ubuntu
+      chpasswd: { expire: False }
+      ssh_pwauth: True
+    - name: root
+      ssh_authorized_keys:
+        - '
+
+cat /dev/zero | ssh-keygen -q -t rsa -f "$pub_key_dir/id_rsa" -N "" || true
+echo "$USER_DATA$(cat "$pub_key_dir/id_rsa.pub")" > "$pub_key_dir/user-data"
+
+cloud-localds "$pub_key_dir/user-data.img" "$pub_key_dir/user-data"
+
+case $PLATFORM in
+amd64)
+  qemu-system-x86_64 \
+    -enable-kvm \
+    -device e1000,netdev=net0 \
+    -netdev user,id=net0,hostfwd=tcp::5555-:22 \
+    -boot c \
+    -pidfile "$PIDFILE" \
+    -m 2G \
+    -drive "file=$IMAGE_FILE,format=qcow2" \
+    -drive "file=$pub_key_dir/user-data.img,format=raw" \
+    -vga none \
+    -nographic &
+  ;;
+arm64)
+  qemu-system-aarch64 \
+    -accel tcg,thread=multi \
+    -machine virt \
+    -cpu cortex-a57 \
+    -device e1000,netdev=net0 \
+    -netdev user,id=net0,hostfwd=tcp::5555-:22 \
+    -boot c \
+    -pidfile "$PIDFILE" \
+    -m 2G \
+    -monitor telnet::45454,server,nowait \
+    -bios .kvm-images/arm64/QEMU_EFI.fd  \
+    -drive "file=$IMAGE_FILE,format=qcow2" \
+    -drive "file=$pub_key_dir/user-data.img,format=raw" \
+    -vga none \
+    -nographic &
+  ;;
+esac
+
+trap finish EXIT
+
+sleep 30
+n=0
+until [ "$n" -ge 15 ]
+do
+   do_ssh "true" && break
+   n=$((n+1))
+   sleep 30
+done
+
+PID="$(cat "$PIDFILE")"
+
+do_ssh "apt-get update && apt-get install -y apt-transport-https debootstrap debian-archive-keyring jq dpkg-dev rsync ca-certificates curl gnupg-agent software-properties-common && mkdir /build"
+do_ssh "curl -fsSL https://download.docker.com/linux/ubuntu/gpg | apt-key add -"
+do_ssh "add-apt-repository \"deb [arch=$PLATFORM] https://download.docker.com/linux/ubuntu \$(lsb_release -cs) stable\""
```

Comment:
thought: this duplicates some things that are done elsewhere. I wonder if copying over the Makefile and using that or something similar would make it easier to maintain?

Reply:
Probably, trying to use `Makefile` to install common requirements now.